diff --git a/pkg/expect/lib/expect.dart b/pkg/expect/lib/expect.dart
index 82e604a..8a891fa 100644
--- a/pkg/expect/lib/expect.dart
+++ b/pkg/expect/lib/expect.dart
@@ -360,11 +360,12 @@ class Expect {
                      [_CheckExceptionFn check = null,
                       String reason = null]) {
     String msg = reason == null ? "" : "($reason)";
-    if (f is! _Nullary) {
-      // Only throws from executing the funtion body should count as throwing.
-      // The failure to even call `f` should throw outside the try/catch.
-      _fail("Expect.throws$msg: Funciton f not callable with zero arguments");
-    }
+    // TODO(ager): Make this work with Fletch.
+// if (f is! _Nullary) {
+//   // Only throws from executing the funtion body should count as throwing.
+//   // The failure to even call `f` should throw outside the try/catch.
+//   _fail("Expect.throws$msg: Funciton f not callable with zero arguments");
+//  }
     try {
       f();
     } catch (e, s) {
